### PR TITLE
adds hard wrapping functionality

### PR DESCRIPTION
**Pull request by bcoe**
In this pull request I've added support for the concept of `hard` vs. `soft` wrapping:

`soft` wrapping allows words that take up more than `cols` to be printed on a single line.

`hard` wrapping does not allow a row to ever contain more than `cols` characters.

This functionality is modeled after @substack's [node-wordwrap](https://github.com/substack/node-wordwrap).

In this pull I've also added code coverage support, which helps me ensure that I'm not being lazy while writing tests.

## How I Tested

1. I wrote a testing script that allowed me to compare substack's wordwrap library, to the ansi wordwrap module. wordwrap is used by `cliui` and `yargs`, and my goal was to make `ansi-wordwrap` a drop in replacement. Here's the script I wrote:

```js
var assert = require('assert');
var stripAnsi = require('strip-ansi');
var chalk = require('chalk');

var fixture = 'The thisisareallylongword quick brown ' + chalk.red('fox jumped over ') + 'the lazy ' + chalk.green('dog and then ran away with the unicorn.');
var wrap = require('./');
var wrap2 = require('wordwrap');

for (var cols = 1; cols < 30; cols++) {
  console.log('cols = ', cols);
  console.log(wrap(fixture, cols, {hard: true}), '\n---');
  console.log(wrap2.hard(cols)(stripAnsi(fixture)), '\n---');
  assert.equal(stripAnsi(wrap(fixture, cols, {hard: true})), wrap2.hard(cols)(stripAnsi(fixture)));
}
```

I used this script to compare both `soft` and `hard` wrapping between the two modules, and managed to get the output pretty close to parity (I actually think that I found a few bugs in wordwrap, that are not present in ansi-wrap related to wrapping close to column boundaries) \o/

screen shot 2015-10-04 at 2 24 08 pm

2. my second test was to `npm link`, `wrap-ansi` and use it as a drop in replacement for `wordwrap`. Both the `yargs` and `cliui` codebases run with the replacement.

reviewers: @dthre, @sindresorhus, @phated, @nexdrew

_note: I started out by trying to tack this functionality onto the existing algorithm, but it turned out to be a hard enough problem that refactoring the algorithm seemed like the better approach._

### CODE REVIEW

**Comment by nexdrew** [comment]
@bcoe Need to rebase and drop commit 4181246 from this branch, I think. :ok_hand:

**Comment by bcoe** [comment]
@nexdrew I purposefully forked off of the code-coverage branch, my plan was to add a few more tests to this refactor to get coverage up a bit.

**Comment by nexdrew** [comment]
Ok, sorry. Ignore me.

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 node_modules
+.nyc_output
```
Comment (Qix-) [inline .gitignore]: Should be removed in lieu of #4.

File: .travis.yml [hunk 2]
```diff
@@ -3,3 +3,4 @@ node_js:
   - 'iojs'
   - '0.12'
   - '0.10'
+after_success: npm run coverage
```
Comment (Qix-) [inline .travis.yml]: Should be removed in lieu of #4.

File: package.json [hunk 3]
```diff
@@ -51,12 +52,13 @@
...
   ],
   "dependencies": {
-    "splice-string": "^1.0.0"
+    "strip-ansi": "^3.0.0"
   },
   "devDependencies": {
     "ava": "0.0.4",
     "chalk": "^1.1.0",
-    "strip-ansi": "^3.0.0",
+    "coveralls": "^2.11.4",
```
Comment (Qix-) [inline package.json]: Should be removed in lieu of #4.

File: package.json [hunk 4]
```diff
@@ -51,12 +52,13 @@
...
+    "nyc": "^3.2.2",
```
Comment (Qix-) [inline package.json]: Should be removed in lieu of #4.

File: package.json [hunk 5]
```diff
@@ -18,7 +18,8 @@
     "node": ">=0.10.0"
   },
   "scripts": {
-    "test": "xo && node test.js"
+    "test": "xo && nyc node test.js",
```
Comment (Qix-) [inline package.json]: Should be removed in lieu of #4.

File: package.json [hunk 6]
```diff
@@ -18,7 +18,8 @@
...
+    "coverage": "xo && nyc --reporter=text-lcov node test.js | coveralls"
```
Comment (Qix-) [inline package.json]: Should be removed in lieu of #4 (both lines).

File: readme.md [hunk 7]
```diff
@@ -1,4 +1,4 @@
-# wrap-ansi [![Build Status](https://travis-ci.org/chalk/wrap-ansi.svg?branch=master)](https://travis-ci.org/chalk/wrap-ansi)
+# wrap-ansi [![Build Status](https://travis-ci.org/chalk/wrap-ansi.svg?branch=master)](https://travis-ci.org/chalk/wrap-ansi) [![Coverage Status](https://coveralls.io/repos/chalk/wrap-ansi/badge.svg?branch=master)](https://coveralls.io/r/chalk/wrap-ansi?branch=master)
```
Comment (Qix-) [inline readme.md]: Should be removed in lieu of #4.

File: readme.md [hunk 8]
```diff
@@ -46,7 +61,7 @@ Number of columns to wrap the text to.
 
 - [slice-ansi](https://github.com/chalk/slice-ansi) - Slice a string with ANSI escape codes
 - [chalk](https://github.com/chalk/chalk) - Terminal string styling done right
-
+- [jsesc](https://github.com/mathiasbynens/jsesc) - Generate ascii only output from unicode strings. This is useful for creating testing fixtures.
```
Comment (Qix-) [inline readme.md]: Why are you adding this link?

File: .gitignore (hunk 1 above)
Comment (nexdrew) [inline .gitignore]: @Qix- I think the intention is to merge #4 first, at which point commit 4181246 will no longer be a part of this PR. If that's acceptable, then we should essentially ignore the changes from that commit here.

File: readme.md [hunk 9]
```diff
@@ -27,7 +27,22 @@ console.log(wrapAnsi(input, 20));
...
+### wrapAnsi(input, columns, opts)
+
+wrap words to the specified column width. By default the wrap is
+`soft`, so long words may extend past the column length.
+
+the following options can be provided:
+
+* `hard`: `boolean`, should hard wrapping be enabled?
+
+### wrapAnsi.hard(input, columns)
```
Comment (nexdrew) [inline readme.md]: Maybe mention that the `.hard()` method and the `{hard: true}` option are synonymous?

File: readme.md (hunk 8 above)
Comment (bcoe) [inline readme.md]: Happy to remove it. I added it because writing fixtures for tests is a pain in the butt without using a tool like `jsesc`.

File: readme.md (hunk 7 above)
Comment (bcoe) [inline readme.md]: This branch is forked off of #4, so once #4 is landed these changes will not be part of this pull -- by forking from the code-coverage branch, I was able to see coverage as I worked on tests for this branch.

File: readme.md (hunk 8 above)
Comment (nexdrew) [inline readme.md]: +1 for keeping it, seems very useful. Developers like me probably wouldn't know about it otherwise.

Could also add a link for `strip-ansi`.

Comment (Qix-) [inline readme.md]: Up to @sindresorhus when it comes to adding links like that. Probably doesn't belong in this pull request either. Submit another PR? :dango:

File: index.js [hunk 10]
```diff
@@ -40,17 +40,24 @@ function wrapAnsi(code) {
...
 }
 
-module.exports = function (str, cols) {
-	var pre = '';
-	var ret = '';
+// calculate the length of words split on ' ', ignoring
+// the extra characters added by ansi escape codes.
+function wordLengths(str) {
+	return stripAnsi(str).split(' ').map(function (s) {
+		return s.length;
```
Comment (sindresorhus) [inline index.js]: Use https://github.com/sindresorhus/string-width here so we handle fullwidth characters correctly.

File: readme.md [hunk 11]
```diff
@@ -27,7 +27,23 @@ console.log(wrapAnsi(input, 20));
...
+
+long words will be broken up so that they do not extend past the column width.
+Equivalent to `{hard: true}`.
+
+### wrapAnsi.soft(input, columns)
```
Comment (sindresorhus) [inline readme.md]: Not a big fan of having both an option and methods for choosing `soft`/`hard`.

File: test.js [hunk 12]
```diff
@@ -7,18 +7,25 @@ var fn = require('./');
 var fixture = 'The quick brown ' + chalk.red('fox jumped over ') + 'the lazy ' + chalk.green('dog and then ran away with the unicorn.');
 
 test(function (t) {
+	var res5 = fn.hard(fixture, 5);
 	var res20 = fn(fixture, 20);
 	var res30 = fn(fixture, 30);
 
-	t.assert(res20 === 'The quick brown\n\u001b[31mfox jumped over \u001b[39mthe\nlazy \u001b[32mdog and then\u001b[39m\n\u001b[32mran away with the\u001b[39m\n\u001b[32municorn.\u001b[39m');
+	t.assert(res20 === 'The quick brown \x1B[31mfox\x1B[39m\n\x1B[31mjumped over \x1B[39mthe lazy\n\x1B[32mdog and then ran\x1B[39m\n\x1B[32maway with the\x1B[39m\n\x1B[32municorn.\x1B[39m');
```
Comment (sindresorhus) [inline test.js]: We prefer the unicode escape `\u001b` for consistency. Can you keep using that?

File: readme.md (hunk 8 above)
Comment (sindresorhus) [inline readme.md]: Why are you not using it programmatically in the tests?

Not really super relevant to this module, but sure, don't mind having it linked.

Description needs some tweaks, though:

> Generate ASCII-only output from Unicode strings. Useful for creating testing fixtures.

File: readme.md [hunk 13]
```diff
@@ -27,7 +27,23 @@ console.log(wrapAnsi(input, 20));
 
 ## API
 
-### wrapAnsi(input, columns)
+### wrapAnsi(input, columns, opts)
+
+wrap words to the specified column width. By default the wrap is
+`soft`, so long words may extend past the column length.
```
Comment (sindresorhus) [inline readme.md]: Don't hardwrap the description here.

**Comment by sindresorhus** [comment]
@dthree Can you review too?

File: readme.md (hunk 11 above)
Comment (Qix-) [inline readme.md]: It's four letters; my vote is on separate methods unless the code is pretty much the same (in fairness I haven't looked that hard at it; getting ready to fly out!).

Comment (sindresorhus) [inline readme.md]: @Qix- They're just wrappers for the `hard` option internally.

Comment (Qix-) [inline readme.md]: Hmm then an option is probably better.

Comment (bcoe) [inline readme.md]: I think I've addressed most of the other review ... how are folks feeling, should we drop the `.hard` and `.soft`? this was modeled after substack's wordwrap API, I don't feel too strongly though -- happy to simply have an opts object.

Comment (nexdrew) [inline readme.md]: I can see how using separate methods could get unwieldy if other options are needed in the future, unless you use a chaining approach like chalk. Probably safer to just use options. :(

File: readme.md (hunk 13 above)
Comment (dthree) [inline readme.md]: Isn't that ironic. :)

**Comment by dthree** [comment]
@sindresorhus yes :)

File: readme.md (hunk 11 above)
Comment (sindresorhus) [inline readme.md]: Ok, let's drop them.

**Comment by dthree** [comment]
This is looking :+1: to me at this point.

**Comment by bcoe** [comment]
@dthree @sindresorhus if you felt like giving me commit access, I'd be happy to land a few more tests on this branch before merging -- I'd like to get coverage a bit closer to 100%, and test all the weird edge-cases of wrapping.

**Comment by dthree** [comment]
@sindresorhus I'm going to merge this to start.

Wouldn't mind giving him access to this repo if you're fine with that, he's put a ton of work into it.

**Comment by sindresorhus** [comment]
Done! Was planning to do so after this PR anyways. Really appriciate the work you've done on this @bcoe :)

**Comment by sindresorhus** [comment]
Looks good to me too.

**Comment by sindresorhus** [comment]
Landed! :)

732852c4-67d8-11e3-90c3-576016e1aac4

**Comment by dthree** [comment]
Haha :) Thanks for all your work @bcoe
